File: python/sglang/kernels/ops/diffusion/triton/ulysses_qkv.py
# SPDX-License-Identifier: Apache-2.0

import torch
import triton
import triton.language as tl


@triton.jit
def _pack_qkv_destination_major_kernel(
    output_ptr,
    q_ptr,
    k_ptr,
    v_ptr,
    total_elements,
    rows,
    local_heads,
    head_size,
    stride_q_row,
    stride_q_head,
    stride_k_row,
    stride_k_head,
    stride_v_row,
    stride_v_head,
    BLOCK_SIZE: tl.constexpr,
):
    offsets = tl.program_id(0) * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
    mask = offsets < total_elements

    dim = offsets % head_size
    head_slot = offsets // head_size
    local_head = head_slot % local_heads
    row_slot = head_slot // local_heads
    row = row_slot % rows
    destination = row_slot // rows
    global_head = destination * local_heads + local_head

    q = tl.load(
        q_ptr + row * stride_q_row + global_head * stride_q_head + dim,
        mask=mask,
    )
    k = tl.load(
        k_ptr + row * stride_k_row + global_head * stride_k_head + dim,
        mask=mask,
    )
    v = tl.load(
        v_ptr + row * stride_v_row + global_head * stride_v_head + dim,
        mask=mask,
    )
    output_base = head_slot * (3 * head_size) + dim
    tl.store(output_ptr + output_base, q, mask=mask)
    tl.store(output_ptr + output_base + head_size, k, mask=mask)
    tl.store(output_ptr + output_base + 2 * head_size, v, mask=mask)


def pack_qkv_destination_major(
    q: torch.Tensor,
    k: torch.Tensor,
    v: torch.Tensor,
    world_size: int,
    out: torch.Tensor | None = None,
) -> torch.Tensor:
    """Pack matching ``[rows, global_heads, head_size]`` Q/K/V tensors."""
    if q.dim() != 3 or q.shape != k.shape or q.shape != v.shape:
        raise ValueError("q, k, and v must have the same 3D shape")
    if not (q.is_cuda and k.is_cuda and v.is_cuda):
        raise ValueError("q, k, and v must be CUDA tensors")
    if not (q.device == k.device == v.device and q.dtype == k.dtype == v.dtype):
        raise ValueError("q, k, and v must have the same device and dtype")
    if q.stride(-1) != 1 or k.stride(-1) != 1 or v.stride(-1) != 1:
        raise ValueError("q, k, and v must be contiguous in head_size")
    if world_size < 1 or q.shape[1] % world_size != 0:
        raise ValueError("world_size must be positive and divide global_heads")

    rows, global_heads, head_size = q.shape
    local_heads = global_heads // world_size
    expected_shape = (world_size, rows, local_heads, 3 * head_size)
    if out is not None:
        if not (
            out.shape == expected_shape
            and out.is_contiguous()
            and out.dtype == q.dtype
            and out.device == q.device
        ):
            raise ValueError(
                "out must be a contiguous tensor with the expected shape, "
                "device, and dtype"
            )
        output = out
    else:
        output = torch.empty(
            expected_shape,
            dtype=q.dtype,
            device=q.device,
        )
    total_elements = rows * global_heads * head_size
    if total_elements == 0:
        return output

    block_size = 1024
    with torch.get_device_module().device(q.device):
        _pack_qkv_destination_major_kernel[(triton.cdiv(total_elements, block_size),)](
            output,
            q,
            k,
            v,
            total_elements,
            rows,
            local_heads,
            head_size,
            q.stride(0),
            q.stride(1),
            k.stride(0),
            k.stride(1),
            v.stride(0),
            v.stride(1),
            BLOCK_SIZE=block_size,
            num_warps=8,
        )
    return output


__all__ = ["pack_qkv_destination_major"]

File: python/sglang/kernels/aot/csrc/cpu/ulysses_qkv.cpp
#include <ATen/ATen.h>
#include <ATen/Dispatch.h>
#include <ATen/Parallel.h>
#include <torch/library.h>

#include <cstring>

/*
 * Pack Q/K/V into destination-major layout for Ulysses all-to-all.
 *
 * Input:
 *   q, k, v:
 *     [rows, global_heads, head_size]
 *
 * Output:
 *   [world_size, rows, local_heads, 3 * head_size]
 *
 * where:
 *   local_heads = global_heads / world_size
 *
 * Mapping:
 *
 *   global_head = destination * local_heads + local_head
 *
 *   output[destination, row, local_head, 0:D]
 *       = q[row, global_head, :]
 *
 *   output[destination, row, local_head, D:2D]
 *       = k[row, global_head, :]
 *
 *   output[destination, row, local_head, 2D:3D]
 *       = v[row, global_head, :]
 *
 */

void pack_qkv_destination_major_cpu(
    const at::Tensor& q, const at::Tensor& k, const at::Tensor& v, int64_t world_size, at::Tensor& output) {
  TORCH_CHECK(
      q.dim() == 3 && k.dim() == 3 && v.dim() == 3 && q.sizes() == k.sizes() && q.sizes() == v.sizes(),
      "q, k, and v must have the same 3D shape");
  TORCH_CHECK(q.device().is_cpu() && k.device().is_cpu() && v.device().is_cpu(), "q, k, and v must be CPU tensors");
  TORCH_CHECK(
      q.scalar_type() == k.scalar_type() && q.scalar_type() == v.scalar_type(), "q, k, and v must have the same dtype");
  TORCH_CHECK(
      q.layout() == at::kStrided && k.layout() == at::kStrided && v.layout() == at::kStrided,
      "q, k, and v must be strided tensors");
  TORCH_CHECK(q.stride(2) == 1 && k.stride(2) == 1 && v.stride(2) == 1, "q, k, and v must be contiguous in head_size");
  TORCH_CHECK(world_size >= 1, "world_size must be positive");

  const int64_t rows = q.size(0);
  const int64_t global_heads = q.size(1);
  const int64_t head_size = q.size(2);
  TORCH_CHECK(global_heads % world_size == 0, "world_size must divide global_heads");
  const int64_t local_heads = global_heads / world_size;
  TORCH_CHECK(output.device().is_cpu(), "output must be a CPU tensor");
  TORCH_CHECK(output.scalar_type() == q.scalar_type(), "output must have the same dtype as q/k/v");
  TORCH_CHECK(
      output.size(0) == world_size && output.size(1) == rows && output.size(2) == local_heads &&
          output.size(3) == 3 * head_size,
      "invalid output shape");
  TORCH_CHECK(output.is_contiguous(), "output must be contiguous");
  if (q.numel() == 0) {
    return;
  }
  const int64_t q_stride_row = q.stride(0);
  const int64_t q_stride_head = q.stride(1);
  const int64_t k_stride_row = k.stride(0);
  const int64_t k_stride_head = k.stride(1);
  const int64_t v_stride_row = v.stride(0);
  const int64_t v_stride_head = v.stride(1);
  AT_DISPATCH_REDUCED_FLOATING_TYPES(q.scalar_type(), "pack_qkv_destination_major_cpu", [&] {
    const scalar_t* q_ptr = q.data_ptr<scalar_t>();
    const scalar_t* k_ptr = k.data_ptr<scalar_t>();
    const scalar_t* v_ptr = v.data_ptr<scalar_t>();
    scalar_t* out_ptr = output.data_ptr<scalar_t>();
    const size_t head_bytes = static_cast<size_t>(head_size) * sizeof(scalar_t);
    const int64_t total_head_slots = rows * global_heads;
    at::parallel_for(0, total_head_slots, 0, [&](int64_t begin, int64_t end) {
      for (int64_t head_slot = begin; head_slot < end; ++head_slot) {
        const int64_t local_head = head_slot % local_heads;
        const int64_t row_slot = head_slot / local_heads;
        const int64_t row = row_slot % rows;
        const int64_t destination = row_slot / rows;
        const int64_t global_head = destination * local_heads + local_head;
        const scalar_t* q_src = q_ptr + row * q_stride_row + global_head * q_stride_head;
        const scalar_t* k_src = k_ptr + row * k_stride_row + global_head * k_stride_head;
        const scalar_t* v_src = v_ptr + row * v_stride_row + global_head * v_stride_head;
        scalar_t* out = out_ptr + head_slot * (3 * head_size);

        // [Q | K | V]
        std::memcpy(out, q_src, head_bytes);
        std::memcpy(out + head_size, k_src, head_bytes);
        std::memcpy(out + 2 * head_size, v_src, head_bytes);
      }
    });
  });
}
